File: research/experiments/sweep/runs/opt-full-python-sonnet45/src/vec_ops.py
"""
Pure vector arithmetic for n-dimensional optimization.
All operations return new lists and never mutate inputs.
"""

import math
from typing import List


def dot(a: List[float], b: List[float]) -> float:
    """Dot product of two vectors."""
    return sum(ai * bi for ai, bi in zip(a, b))


def norm(v: List[float]) -> float:
    """Euclidean (L2) norm."""
    return math.sqrt(sum(vi * vi for vi in v))


def norm_inf(v: List[float]) -> float:
    """Infinity norm (max absolute value)."""
    if not v:
        return 0.0
    return max(abs(vi) for vi in v)


def scale(v: List[float], s: float) -> List[float]:
    """Scalar multiplication."""
    return [vi * s for vi in v]


def add(a: List[float], b: List[float]) -> List[float]:
    """Element-wise addition."""
    return [ai + bi for ai, bi in zip(a, b)]


def sub(a: List[float], b: List[float]) -> List[float]:
    """Element-wise subtraction."""
    return [ai - bi for ai, bi in zip(a, b)]


def negate(v: List[float]) -> List[float]:
    """Element-wise negation."""
    return [-vi for vi in v]


def clone(v: List[float]) -> List[float]:
    """Deep copy."""
    return v[:]


def zeros(n: int) -> List[float]:
    """Vector of n zeros."""
    return [0.0] * n


def add_scaled(a: List[float], b: List[float], s: float) -> List[float]:
    """a + s*b (fused, avoids intermediate allocation)."""
    return [ai + s * bi for ai, bi in zip(a, b)]

File: research/experiments/sweep/runs/opt-full-rust-sonnet45/src/vec_ops.rs
/// Pure vector arithmetic for n-dimensional optimization.
/// All operations return new vectors and never mutate inputs.

pub fn dot(a: &[f64], b: &[f64]) -> f64 {
    a.iter().zip(b.iter()).map(|(ai, bi)| ai * bi).sum()
}

pub fn norm(v: &[f64]) -> f64 {
    dot(v, v).sqrt()
}

pub fn norm_inf(v: &[f64]) -> f64 {
    v.iter().map(|x| x.abs()).fold(0.0_f64, f64::max)
}

pub fn scale(v: &[f64], s: f64) -> Vec<f64> {
    v.iter().map(|x| x * s).collect()
}

pub fn add(a: &[f64], b: &[f64]) -> Vec<f64> {
    a.iter().zip(b.iter()).map(|(ai, bi)| ai + bi).collect()
}

pub fn sub(a: &[f64], b: &[f64]) -> Vec<f64> {
    a.iter().zip(b.iter()).map(|(ai, bi)| ai - bi).collect()
}

pub fn negate(v: &[f64]) -> Vec<f64> {
    v.iter().map(|x| -x).collect()
}

pub fn clone_vec(v: &[f64]) -> Vec<f64> {
    v.to_vec()
}

pub fn zeros(n: usize) -> Vec<f64> {
    vec![0.0; n]
}

pub fn add_scaled(a: &[f64], b: &[f64], s: f64) -> Vec<f64> {
    a.iter().zip(b.iter()).map(|(ai, bi)| ai + s * bi).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_dot() {
        assert_eq!(dot(&[1.0, 2.0, 3.0], &[4.0, 5.0, 6.0]), 32.0);
        assert_eq!(dot(&[0.0, 0.0], &[1.0, 1.0]), 0.0);
    }

    #[test]
    fn test_norm() {
        assert_eq!(norm(&[3.0, 4.0]), 5.0);
        assert_eq!(norm(&[0.0, 0.0, 0.0]), 0.0);
    }

    #[test]
    fn test_norm_inf() {
        assert_eq!(norm_inf(&[1.0, -3.0, 2.0]), 3.0);
        assert_eq!(norm_inf(&[0.0, 0.0]), 0.0);
    }

    #[test]
    fn test_scale() {
        assert_eq!(scale(&[1.0, 2.0], 3.0), vec![3.0, 6.0]);
        assert_eq!(scale(&[1.0, 2.0], 0.0), vec![0.0, 0.0]);
    }

    #[test]
    fn test_add() {
        assert_eq!(add(&[1.0, 2.0], &[3.0, 4.0]), vec![4.0, 6.0]);
    }

    #[test]
    fn test_sub() {
        assert_eq!(sub(&[3.0, 4.0], &[1.0, 2.0]), vec![2.0, 2.0]);
    }

    #[test]
    fn test_negate() {
        assert_eq!(negate(&[1.0, -2.0]), vec![-1.0, 2.0]);
    }

    #[test]
    fn test_clone() {
        let v = vec![1.0, 2.0];
        let c = clone_vec(&v);
        assert_eq!(c, vec![1.0, 2.0]);
        // Distinct: modifying clone doesn't affect original
        let mut c2 = clone_vec(&v);
        c2[0] = 99.0;
        assert_eq!(v[0], 1.0);
    }

    #[test]
    fn test_zeros() {
        assert_eq!(zeros(3), vec![0.0, 0.0, 0.0]);
    }

    #[test]
    fn test_add_scaled() {
        assert_eq!(add_scaled(&[1.0, 2.0], &[3.0, 4.0], 2.0), vec![7.0, 10.0]);
    }

    #[test]
    fn test_purity() {
        let a = vec![1.0, 2.0];
        let b = vec![3.0, 4.0];
        let _ = add(&a, &b);
        assert_eq!(a, vec![1.0, 2.0]);
        assert_eq!(b, vec![3.0, 4.0]);

        let v = vec![1.0, 2.0];
        let _ = scale(&v, 5.0);
        assert_eq!(v, vec![1.0, 2.0]);
    }
}
